### ISSUE TITLE
Allow stable ABI extensions to include multiarch tuples in the filename

### ISSUE THREAD

**Issue by stefanor** [python/cpython#122931]
# Feature or enhancement

### Proposal:

Currently C extensions can include the multarch tuple of the architecture the support in their filename, e.g. `zlib.cpython-314-x86_64-linux-gnu.so`. This extensions for multiple different architectures to be installed on the same filesystem, next to each other, without clashing. This is used  [by Debian](https://wiki.debian.org/Multiarch) to allow packages for multiple architectures to be co-installed simultaneously.

The same should apply to stable ABI C extensions, but this was missed when the stable ABI was implemented.

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

### Linked PRs
* gh-122917
* gh-152461
* gh-152831
* gh-154772

**Comment by rgommers** [comment]
This will not be so easy to introduce. Currently build systems/backends get the `cpython-314-x86_64-linux-gnu.so` extension name from a value like `sysconfig.get_config_var('EXT_SUFFIX')` while they hardcode `abi3.so` because it's the only option. 

I assume you want all build tools to do the new thing and produce `abi3-x86_64-linux-gnu.so` always instead of `abi3.so`? There are a lot of build tools to update then.

**Comment by stefanor** [comment]
> This will not be so easy to introduce.

The plan was: Start supporting this as an alternative, until we reach a point where we don't have to support older versions that don't support it any more. Then it can become the default.

We have supported this in Debian since 3.13, and were hoping to get support upstream into 3.14.0, but moved a little too late for that :(

> I assume you want all build tools to do the new thing and produce abi3-x86_64-linux-gnu.so always instead of abi3.so? There are a lot of build tools to update then.

If build tools are asking the interpreter what to prefer, then they should do the right thing. Which is to not build anything with this, for now.

We handle renaming extensions to include this path in our package tooling. Nothing else really needs it, at this point.

**Comment by rgommers** [comment]
> We handle renaming extensions to include this path in our package tooling. Nothing else really needs it, at this point.

Ah great, thanks for the context. No concerns from me then.

**Comment by eli-schwartz** [comment]
> This will not be so easy to introduce. Currently build systems/backends get the `cpython-314-x86_64-linux-gnu.so` extension name from a value like `sysconfig.get_config_var('EXT_SUFFIX')` while they hardcode `abi3.so` because it's the only option.

This is of course not true. Meson uses `importlib.machinery.EXTENSION_SUFFIXES[1]` on the theory that element 0 is the the EXT_SUFFIX (because it is most specific and thus most preferred) while element 1 is the second-most preferred which ought to be the stable ABI suffix.

Inserting another abi3 suffix so that it looks like

```
['.cpython-314-x86_64-linux-gnu.so', '.abi3-x86_64-linux-gnu.so', '.abi3.so', '.so']
```

is:
- intrinsically necessary (to include in the list) to get the importlib machinery to work correctly
- intrinsically necessary (to use that specific order) because an architecture specific abi3 extension needs to take priority over an abi3 extension of unknown compatibility with the current interpreter
- automatically going to work for meson


> The plan was: Start supporting this as an alternative, until we reach a point where we don't have to support older versions that don't support it any more. Then it can become the default.

This is an unnecessary complication since both names have to be supported for compatibility, so the default can immediately become the "better" one. Tools that don't know how to use the better one don't *have* to, because for compatibility they both work! Tools that do.know how to use the better one... should do so?

Older python versions don't matter, because the official word from on high is that if you compile an abi3 extension on python 3.13, you're not allowed to use it with a python 3.12 interpreter. Rationale is that the minimum limited API setting being lower than the current headers version, "doesn't actually guarantee you're avoiding stable ABIs from the most recent version", and shall not be supported, therefore it's useless to say you'd like to be able to do it.

**Comment by stefanor** [comment]
> Meson uses importlib.machinery.EXTENSION_SUFFIXES[1]

That does sound a little fragile. It select take the first one with `.abi` in it, and be a little more resilient to change. setuptools does that:
https://github.com/pypa/setuptools/blob/d198e86f57231e83de87975c5c82bc40c196da79/setuptools/command/build_ext.py#L79-L86

> intrinsically necessary (to use that specific order) because an architecture specific abi3 extension needs to take priority over an abi3 extension of unknown compatibility with the current interpreter

I don't think we want that behaviour until we can assume that any interpreter that is able to get this extension in a wheel can import this filename.

> This is an unnecessary complication since both names have to be supported for compatibility, so the default can immediately become the "better" one.

Same issue. If you're building any binaries that other people are going to use with other interpreters, you need to ensure that they can import them.

**Comment by eli-schwartz** [comment]
> Older python versions don't matter, because the official word from on high is that if you compile an abi3 extension on python 3.13, you're not allowed to use it with a python 3.12 interpreter.

**Comment by eli-schwartz** [comment]
> > Meson uses importlib.machinery.EXTENSION_SUFFIXES[1]
> 
> That does sound a little fragile.

It is! It doesn't work in cross compilation! Changing this would be lovely, but playing games with modifying the heuristics about how to parse details of a runtime import of the importlib machinery are very uninteresting.

In fact it's already been changed -- if you use a PEP 739 build_details.json, the abi.stable_abi_suffix is mandatory for CPython (and any other python implementations that have a stable ABI) and is/can be used in preference to guessing anything.

One would presume therefore that that build_details.json would list the intended ABI suffix (the one that is most specific to the current interpreter, i.e. the one that lists the CPU arch).

**Comment by stefanor** [comment]
>  Older python versions don't matter, because the official word from on high is that if you compile an abi3 extension on python 3.13, you're not allowed to use it with a python 3.12 interpreter.

IIRC that's a new thing, since we made this proposal. And there's nothing to declare the compatible versions in the data...

Also, right now if you compile on a *Debian* Python 3.13, the multiarch extension is there...

**Comment by eli-schwartz** [comment]
> And there's nothing to declare the compatible versions in the data...

@zooba specifically wrote PEP 809 to address this very point -- the stable ABI would permit neither additions nor removals except when revbumping the ABI tag (abi4, abi2026, depending).



> Also, right now if you compile on a _Debian_ Python 3.13, the multiarch extension is there...

If the idea is to add this change to the official unpatched interpreter, this problem doesn't exist.

**Comment by merwok** [comment]
> > Older python versions don't matter, because the official word from on high is that if you compile an abi3 extension on python 3.13, you're not allowed to use it with a python 3.12 interpreter.
> 
> IIRC that's a new thing, since we made this proposal.

No, that’s how the stable ABI works: https://docs.python.org/3/c-api/stable.html#limited-c-api

You can compile with limited API version 3.12 using Python 3.12 or 3.13, then it will work on 3.12 and higher. Or you compile with limited API version 3.13 on 3.13, and it won’t load on 3.12 and older.

**Comment by rgommers** [comment]
> This is of course not true. Meson uses `importlib.machinery.EXTENSION_SUFFIXES[1]`

Thanks for the correction. That makes it more fragile overall; `meson-python` does do what I described. `sysconfig.get_config_var('EXT_SUFFIX')` is only a single string value - hard to modify in a backwards compatible way.

> Older python versions don't matter, because the official word from on high is that if you compile an abi3 extension on python 3.13, you're not allowed to use it with a python 3.12 interpreter.

I don't think I've ever seen that in writing. _Using_ a 3.13 interpreter with `-DPy_LIMITED_API=<3.12 hex version>` in compile flags is common and should work. I'm guessing you meant `-DPy_LIMITED_API=<3.13 hex version>`?

**Comment by eli-schwartz** [comment]
> I don't think I've ever seen that in writing. _Using_ a 3.13 interpreter with `-DPy_LIMITED_API=<3.12 hex version>` in compile flags is common and should work. I'm guessing you meant `-DPy_LIMITED_API=<3.13 hex version>`?

I'm saying that I am pretty sure I've seen CPython core developers say all the time that you may use whatever Py_LIMITED_API you like, and it has the effect of fixing a lower bound, but it doesn't set an upper bound and it doesn't guarantee that the CPython headers won't silently replace functions with macros that use *newer* ABI than what you expect.

Therefore the only guarantee we have as users is that if you compile an extension using a Py_LIMITED_API specifying version X, it will:
- error out if CPython is X-1
- for CPython >= X, produces a compiled extension that requires at least $CPYTHON_VERSION and pays zero attention to Py_LIMITED_API

Or in simpler terms, specifying a particular version for Py_LIMITED_API does nothing except provide partial linting for cases where your source code is incompatible with the claim that Py_LIMITED_API is making. And lint tools do not an ABI make.

It is unfortunate that no available mechanisms are made use of *in the CPython codebase* for offering better compile time or runtime guarantees (runtime guarantees for this exact scenario were pioneered decades ago and heavily used by almost every Linux binary on the planet because glibc injects those guarantees whether you know it does or not, every time a compiled binary uses functions from glibc, which is ~100% of them, -+ 0%).

**Comment by stefanor** [comment]
@eli-schwartz: I don't disagree that it's defined that stable ABI doesn't work backords, or that PEP 809 proposes a future that makes this stuff clearer.

But right now, it's very hard to build tooling that respects those rules, because the tools don't have the data. So we have to be conservative in our approach to change, to not publish binaries that break existing systems. Tools like auditwheel are a great help here, but I'm still trying to be careful.

**Comment by eli-schwartz** [comment]
> But right now, it's very hard to build tooling that respects those rules, because the tools don't have the data. So we have to be conservative in our approach to change, to not publish binaries that break existing systems. Tools like auditwheel are a great help here, but I'm still trying to be careful.

But the whole point of the change is that it:
- only affects new systems, not existing ones
- produces binaries that are less capable of breaking systems than the current state of the art

Giving more data to tools sounds like the opposite of a problem.

And it's precisely the fact that the ABI doesn't work backwards, that prevents this from being an issue. Any tool that doesn't understand the change -- and CPython 3.12 is just such a tool -- has to ignore it and pretend it doesn't exist, which is exactly what you want anyway as the alternative is to load an ABI-incompatible module with unpredictable failure modes.

**Comment by stefanor** [comment]
Made some progress on pushing this forward at PyCon US.

On a Linux systems, we currently support importing C extensions from a few possible extensions. Merging the fix to #122931 possibly brings this list to 6:

```python 
>>> pprint.pprint(_imp.extension_suffixes())
[
    '.cpython-316-x86_64-linux-gnu.so',  # The default suffix generated by build tools
    '.abi3-x86_64-linux-gnu.so',  # Introduced here
    '.abi3.so',  # Legacy - must be supported as long as possible
    '.abi3t-x86_64-linux-gnu.so',  # Introduced here
    '.abi3t.so',  # New in 3.15. To be deprecated. Can we skip it entirely?
    '.so',  # Legacy
]
```

If we push forward with this fast, we can avoid ever having to support `.abi3t.so`.

I did a review of build tools to see what the impact will be. PyPA build tools currently produce the fully-specified (1st) option, unless they are building stable ABI extensions. With this change, we expect stable ABI builds to quickly start producing to fully-specified abi3 files (2nd), as they are using modern build tooling.

Tools used to build C extensions in at least 2 packages in the [top 15k](https://github.com/hugovk/top-pypi-packages):

```
      2  _build_backend
      2  _pyyaml_pep517
      2  builder
      2  enscons.api
      2  ext
      2  nuitka.distutils.Build
      2  scikit_build_core_conan.build
      2  setup.build_meta
      3  flit_core.buildapi
      3  scikit_build_core.setuptools.build_meta
      4  build_backend
      4  pep517_backend.hooks
      4  pup517.backend
      4  xmake_python
      5  packager.pep517
      5  setup
      6  pdm.pep517.api
      7  uv_build
     10  poetry_dynamic_versioning.backend
     11  nvidia_stub.buildapi
     11  wheel_stub.buildapi
     13  py_build_cmake.build
     17  poetry.masonry.api
     23  sipbuild.api
     26  backend
     28  pdm.backend
     52  cmeel
    145  hatchling.build
    191  mesonpy
    245  poetry.core.masonry.api
    646  scikit_build_core.build
   2163  ERROR  # Couldn't determine a tool for some reason. e.g. no source package or invalid pyproject.toml
   2843  maturin
   2861  setuptools.build_meta
```

Tool support:
* setuptools: [selects the first `abi3` extension in `extension_suffixes`](https://github.com/pypa/setuptools/blob/84ed5913724df5a12dc804e1d5efe12508e706d2/setuptools/command/build_ext.py#L79-L86)
* setuptools-rust uses the same code
* meson-python: [selects the 2nd item in `extension_suffixes`](https://github.com/mesonbuild/meson/blob/d8af51917cbadc3c9bd53f42b72e9486e3e9d6b0/mesonbuild/scripts/python_info.py#L95-L101) (TODO: room for improvement)
* maturin: [generates its extension name directly](https://github.com/PyO3/maturin/blob/7543032bbc9e510e340dea162b3a4e5295eada7d/src/binding_generator/pyo3_binding.rs#L80) - will need to be fixed
* cmake / scikit-build-core: [also generates its extension name](https://github.com/Kitware/CMake/blob/24d023247a7f67bf74f624b5a37b34183c570bf4/Modules/FindPython/Support.cmake#L557-L567) - will need to be fixed

Some of these tools (`maturin` and `cmake`) already support `abi3t` and will need to learn to do it better.

Tools building `abi3` for older releases with newer Pythons will need to be updated to know not to use the multiarch tags. It would be safe to backport support for the multiarch tags at a lower priority than the un-tagged version), which would mitigate this.

We can probably immediately deprecate `.abi3t.so`, to the point that it never needs to be supported in a supported release of Python.

**Comment by eli-schwartz** [comment]
> * meson-python: [selects the 2nd item in `extension_suffixes`](https://github.com/mesonbuild/meson/blob/d8af51917cbadc3c9bd53f42b72e9486e3e9d6b0/mesonbuild/scripts/python_info.py#L95-L101) (TODO: room for improvement)

With the caveat that this is the "non build_details.json" path, yes.

**Comment by rgommers** [comment]
This is now completely implemented and can be closed, right? gh-152461 was the last part of this, and that was merged and backported before 3.15.0rc1.
